File: solutions/3018-make-string-a-subsequence-using-cyclic-increments/make-string-a-subsequence-using-cyclic-increments.rb
def can_make_subsequence(source, target)
    src_len = source.length
    tgt_len = target.length
    target_char = target[0]
    
    src_idx = 0
    tgt_idx = 0
    
    while src_idx < src_len && tgt_idx < tgt_len
        src_char = source[src_idx]
        
        if src_char == target_char || 
           src_char.next[0] == target_char || 
           (src_char == 'z' && target_char == 'a')
            tgt_idx += 1
            target_char = target[tgt_idx] if tgt_idx < tgt_len
        end
        
        src_idx += 1
    end
    
    tgt_idx == tgt_len
end

File: solutions/3018-make-string-a-subsequence-using-cyclic-increments/make-string-a-subsequence-using-cyclic-increments.js
// JavaScript

var canMakeSubsequence = function(source, target) {
    const srcLen = source.length;
    const tgtLen = target.length;
    let targetChar = target[0];
    
    let srcIdx = 0, tgtIdx = 0;
    while (srcIdx < srcLen && tgtIdx < tgtLen) {
        const srcChar = source[srcIdx];
        
        if (srcChar === targetChar || 
            String.fromCharCode(srcChar.charCodeAt(0) + 1) === targetChar ||
            (srcChar === 'z' && targetChar === 'a')) {
            tgtIdx++;
            if (tgtIdx < tgtLen) {
                targetChar = target[tgtIdx];
            }
        }
        
        srcIdx++;
    }
    
    return tgtIdx === tgtLen;
};